Mathlib/AlgebraicGeometry/ValuativeCriterion.lean

```diff
@@ -0,0 +1,321 @@
+Authors: Andrew Yang, Qi Ge, Christian Merten
+-/
+import Mathlib.AlgebraicGeometry.Morphisms.Immersion
+import Mathlib.AlgebraicGeometry.Morphisms.Proper
+import Mathlib.RingTheory.RingHom.Injective
+import Mathlib.RingTheory.Valuation.LocalSubring
+
+/-!
+# Valuative criterion
+
+## Main results
+
+- `AlgebraicGeometry.UniversallyClosed.eq_valuativeCriterion`:
+  A morphism is universally closed if and only if
+  it is quasi-compact and satisfies the existence part of the valuative criterion.
+- `AlgebraicGeometry.UniversallyClosed.eq_valuativeCriterion`:
+  A morphism is separated if and only if
+  it is quasi-separated and satisfies the uniqueness part of the valuative criterion.
+- `AlgebraicGeometry.UniversallyClosed.eq_valuativeCriterion`:
+  A morphism is proper if and only if
+  it is qcqs and of fintite type and satisfies the valuative criterion.
+
+## Future projects
+Show that it suffices to check discrete valuation rings when the base is noetherian.
+
+-/
+
+open CategoryTheory CategoryTheory.Limits
+
+namespace AlgebraicGeometry
+
+universe u
+
+/--
+A valuative commutative square over a morphism `f : X ⟶ Y` is a square
+```
+Spec K ⟶ Y
+  |       |
+  ↓       ↓
+Spec R ⟶ X
+```
+where `R` is a valuation ring, and `K` is its ring of fractions.
+
+We are interested in finding lifts `Spec R ⟶ Y` of this diagram.
+-/
+structure ValuativeCommSq {X Y : Scheme.{u}} (f : X ⟶ Y) where
+  /-- The valuation ring of a valuative commutative square. -/
+  R : Type u
+  [commRing : CommRing R]
+  [domain : IsDomain R]
+  [valuationRing : ValuationRing R]
+  /-- The field of fractions of a valuative commutative square. -/
+  K : Type u
+  [field : Field K]
+  [algebra : Algebra R K]
+  [isFractionRing : IsFractionRing R K]
+  /-- The top map in a valuative commutative map. -/
+  (i₁ : Spec (.of K) ⟶ X)
+  /-- The bottom map in a valuative commutative map. -/
+  (i₂ : Spec (.of R) ⟶ Y)
+  (commSq : CommSq i₁ (Spec.map (CommRingCat.ofHom (algebraMap R K))) f i₂)
+
+namespace ValuativeCommSq
+
+attribute [instance] commRing domain valuationRing field algebra isFractionRing
+
+end ValuativeCommSq
+
+/-- A morphism `f : X ⟶ Y` satisfies the existence part of the valuative criterion if
+every valuative commutative square over `f` has (at least) a lift. -/
+def ValuativeCriterion.Existence : MorphismProperty Scheme :=
+  fun _ _ f ↦ ∀ S : ValuativeCommSq f, S.commSq.HasLift
+
+/-- A morphism `f : X ⟶ Y` satisfies the uniqueness part of the valuative criterion if
+every valuative commutative square over `f` has at most one lift. -/
+def ValuativeCriterion.Uniqueness : MorphismProperty Scheme :=
+  fun _ _ f ↦ ∀ S : ValuativeCommSq f, Subsingleton S.commSq.LiftStruct
+
+/-- A morphism `f : X ⟶ Y` satisfies the valuative criterion if
+every valuative commutative square over `f` has a unique lift. -/
+def ValuativeCriterion : MorphismProperty Scheme :=
+  fun _ _ f ↦ ∀ S : ValuativeCommSq f, Nonempty (Unique (S.commSq.LiftStruct))
```

Comment:
Please add a little bit of API:
- projections `.existence` and `.uniqueness`,
- and `iff` that shows this is equivalent to `Existence + Uniqueness`.

Reply:
Done.